### PR TITLE
ci: add CodeQL analysis

### PR DESCRIPTION
**Pull request by twpayne**
This PR enables [CodeQL analysis](https://securitylab.github.com/tools/codeql) for the project.

cc @sharlns

### CODE REVIEW

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    steps:
+    - name: Checkout repo
+      uses: actions/checkout@v2
+      with:
+        fetch-depth: 1
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v1
+      with:
+        languages: go
+    - name: Perform CodeQL Analysis
```
Comment (sayboras) [inline .github/workflows/codeql-analysis.yml:30]: do we need to add step to do either autobuild or manual build before this step ?

File: .github/workflows/codeql-analysis.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    # - master
+  pull_request:
+    # branches:
+    # - master
+  # schedule:
+  #   - cron: "0 3 * * 4"
+
+jobs:
+  analyze:
+    runs-on: ubuntu-18.04
```
Comment (sayboras) [inline .github/workflows/codeql-analysis.yml:18]: maybe 20.04 or even latest.

**Review by sayboras — COMMENTED** [review]
Nice 💯, one quick question on report https://github.com/cilium/cilium/security/code-scanning?query=ref%3Arefs%2Fheads%2Fpr%2Ftwpayne%2Fcodeql-analysis, just curious who will check this report and take any action if required ? 🤔

File: .github/workflows/codeql-analysis.yml (hunk 1 above)
Comment (twpayne) [reply .github/workflows/codeql-analysis.yml:30]: I don't think it's necessary. CodeQL does source code analysis only, it doesn't use any build artifacts. IIUC All the generated source code is already checked into Cilium's repo so no build steps are required.

File: .github/workflows/codeql-analysis.yml (hunk 2 above)
Comment (twpayne) [reply .github/workflows/codeql-analysis.yml:18]: TBH I don't think it matters. I chose 18.04 as it generally has wider support (still) and avoids the "`ubuntu-latest` will soon default to `ubuntu-20.04`" warnings in GitHub Actions.

**Comment by twpayne** [comment]
> just curious who will check this report and take any action if required ?

Initially me and probably @sharlns 😄 We're both working for Isovalent and have an interest in Cilium's security.

**Review by sayboras — APPROVED** [review]
LGTM 💯

File: .github/workflows/codeql-analysis.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+# FIXME uncomment conditions before merging PR
+on:
+  push:
+    # branches:
+    # - master
+  pull_request:
+    # branches:
+    # - master
+  # schedule:
+  #   - cron: "0 3 * * 4"
```
Comment (kkourt) [inline .github/workflows/codeql-analysis.yml]: Is the intention to uncomment the above cron line? If yes, why is it not sufficient to run this just on PRs?

**Review by aanm — COMMENTED** [review]
Any reason why we don't enable it by default as a blocking requirement to merge PRs? How long does the codeql takes to run? It looked it was less than 3 seconds?

File: .github/workflows/codeql-analysis.yml (hunk 3 above)
Comment (twpayne) [reply .github/workflows/codeql-analysis.yml]: Yes. The CodeQL scanner is evolving and gets new checks added to it from time to time. So, even if the code hasn't changed, the scanner might have done, so we should re-run the scanner.

**Comment by twpayne** [comment]
> Any reason why we don't enable it by default as a blocking requirement to merge PRs?

The scanner does occasionally have false positives (e.g. https://github.com/github/codeql-go/issues/439) and not all the problems it identifies are genuine security problems so I think making it a blocking requirement would be too strict.

**Review by kkourt — APPROVED** [review]
Thanks.

**Comment by twpayne** [comment]
I think this can be merged now. Many of the identified problems have been fixed in other PRs which can be merged one by one.

File: .github/workflows/codeql-analysis.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    branches:
+    - master
+    - v1.9
+    - v1.8
+    - v1.7
+  pull_request:
+    branches:
+    - master
+  schedule:
+    - cron: "6 45 * * 3"
```
Comment (pchaigno) [inline .github/workflows/codeql-analysis.yml:14]: @twpayne Was there a particular reason for wanting to run this both on a schedule *and* on every push to master?

File: .github/workflows/codeql-analysis.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+jobs:
+  analyze:
+    runs-on: ubuntu-18.04
+    strategy:
+      fail-fast: false
```
Comment (pchaigno) [inline .github/workflows/codeql-analysis.yml:20]: @twpayne Why is this necessary? It seems like each step in this job depends on the previous steps anyway, no?

File: .github/workflows/codeql-analysis.yml (hunk 4 above)
Comment (twpayne) [reply .github/workflows/codeql-analysis.yml:14]: Yes, the CodeQL service gets updated from time to time as they add more checks, so we want to take advantage of those.

Comment (pchaigno) [reply .github/workflows/codeql-analysis.yml:14]: It feels a bit excessive maybe. This job takes between 17 and 25min to execute, and during that time, block the slot for another job. Would it be enough to run it on all PRs + once a day, but not on every commit?

Maybe I'm underestimating how many bugs it found so far? I also had the impression it's a static analysis tool, so running it on every single commit doesn't give us more information than running on a schedule because it's already clear what is broken when the test fails. Am I missing something?

Comment (twpayne) [reply .github/workflows/codeql-analysis.yml:14]: I don't think you're missing anything, your analysis sounds very reasonable.

It found a few potential bugs in the initial analysis, but few since. It's nice to get feedback early while the PR is still being reviewed as this allows the PR author and reviewers to solve problems before the PR is merged, rather than having to follow-up afterwards.

That all said, if it's taking a slot from other jobs, then I agree that it's worth reducing the frequency at which CodeQL is run so we can speed up CI for individual PRs.

File: .github/workflows/codeql-analysis.yml (hunk 5 above)
Comment (twpayne) [reply .github/workflows/codeql-analysis.yml:20]: Is `fail-fast` only specific to a single GitHub workflow? From (possibly fallible) memory, at the time this PR was being developed, CI was very flaky, and CI flakes were causing the CodeQL analysis to be cancelled. I added this so that the analysis would run to completion.

File: .github/workflows/codeql-analysis.yml (hunk 4 above)
Comment (pchaigno) [reply .github/workflows/codeql-analysis.yml:14]: > It's nice to get feedback early while the PR is still being reviewed as this allows the PR author and reviewers to solve problems before the PR is merged, rather than having to follow-up afterwards.

I can't review all pull requests when they are opened. Sometimes we find things after they are merged, while looking for something else. I just wanted to discuss these concerns in the context of the code. In any case, whenever I'm raising concerns and we can agree, I'm sending the fixes myself.

Comment (pchaigno) [reply .github/workflows/codeql-analysis.yml:14]: I've opened https://github.com/cilium/cilium/pull/16241 for this. Thanks for taking the time to answer @twpayne! :pray:

File: .github/workflows/codeql-analysis.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+name: codeql
+
+on:
+  push:
+    branches:
+    - master
+    - v1.9
```
Comment (michi-covalent) [inline .github/workflows/codeql-analysis.yml:7]: late to the party, but maybe `v*.*` here? seems pretty easy to forget to update this file after each release.

Comment (pchaigno) [reply .github/workflows/codeql-analysis.yml:7]: :ship:

Comment (michi-covalent) [reply .github/workflows/codeql-analysis.yml:7]: https://github.com/cilium/cilium/pull/16251 🚢 

also removed the periodic trigger. let me know if we still want to keep it.
